File: solutions/go/spiral-matrix/1/spiral_matrix.go
package spiralmatrix

func SpiralMatrix(size int) [][]int {
	// Handle edge case.
	if size == 0 {
		return [][]int{}
	}

	matrix := make([][]int, size)

	for i := range matrix {
		matrix[i] = make([]int, size)
	}

	// Moving right example
	row := 0
	col := 0

	current := 1

	top := 0
	bottom := size - 1
	left := 0
	right := size - 1

	for current <= size*size {
		// Move right
		for col <= right && current <= size*size {
			matrix[row][col] = current
			current++
			col++
		}
		col-- // Fix position
		top++ // Update boundary
		row++ // Prepare for next direction

		// Move down
		for row <= bottom && current <= size*size {
			matrix[row][col] = current
			current++
			row++
		}
		row--   // Fix position
		right-- // Update boundary
		col--   // Prepare for next direction

		// Move left
		for col >= left && current <= size*size {
			matrix[row][col] = current
			current++
			col--
		}
		col++    // Fix position
		bottom-- // Update boundary
		row--    // Prepare for next direction

		// Move up
		for row >= top && current <= size*size {
			matrix[row][col] = current
			current++
			row--
		}
		row++  // Fix position
		left++ // Update boundary
		col++  // Prepare for next direction
	}

	return matrix

}

File: solutions/typescript/spiral-matrix/1/spiral-matrix.ts
export function ofSize(input: number): number[][] {
  let matrix: number[][] = []
  let counter = 1
  let row = 0
  let col = 0
  let size = input
  let direction = 'right'

  for (let i = 0; i < size; i++) {
    matrix.push([])
    for (let j = 0; j < size; j++) {
      matrix[i].push(0)
    }
  }

  while (counter <= size * size) {
    matrix[row][col] = counter

    if (direction === 'right') {
      if (col + 1 < size && matrix[row][col + 1] === 0) {
        col++
      } else {
        direction = 'down'
        row++
      }
    } else if (direction === 'down') {
      if (row + 1 < size && matrix[row + 1][col] === 0) {
        row++
      } else {
        direction = 'left'
        col--
      }
    } else if (direction === 'left') {
      if (col - 1 >= 0 && matrix[row][col - 1] === 0) {
        col--
      } else {
        direction = 'up'
        row--
      }
    } else if (direction === 'up') {
      if (row - 1 >= 0 && matrix[row - 1][col] === 0) {
        row--
      } else {
        direction = 'right'
        col++
      }
    }

    counter++
  }

  return matrix
}
